tests/specs/lemmas.k

```diff
@@ -74,17 +74,58 @@ module LEMMAS
 
     rule nthbyteof(N, 0, 1) => N
 
+    rule WS ++ .WordStack => WS
 
+    rule #sizeWordStack(#buf(N, _)  , M) => M +Int N
+    rule #sizeWordStack(BUF1 ++ BUF2, M) => M +Int #sizeWordStack(BUF1) +Int #sizeWordStack(BUF2)
+    rule #sizeWordStack(#drop(N, BUF))   => maxInt(#sizeWordStack(BUF) -Int N, 0)
 
+    rule #take(N, BUF)          => BUF                                              requires N ==Int #sizeWordStack(BUF)
+    rule #take(N, BUF1 ++ BUF2) => #take(N, BUF1)                                   requires N <=Int #sizeWordStack(BUF1)
+    rule #take(N, BUF1 ++ BUF2) => BUF1 ++ #take(N -Int #sizeWordStack(BUF1), BUF2) requires N  >Int #sizeWordStack(BUF1)
 
+    rule #drop(N, BUF)           => .WordStack                               requires #sizeWordStack(BUF) <=Int N
+    rule #drop(N, #drop(M, BUF)) => #drop(N +Int M, BUF)
+    rule #drop(N, BUF1 ++ BUF2)  => #drop(N -Int #sizeWordStack(BUF1), BUF2) requires N >=Int #sizeWordStack(BUF1)
+    rule #drop(N, BUF1 ++ BUF2)  => #drop(N, BUF1) ++ BUF2                   requires N <Int #sizeWordStack(BUF1)
 
+    rule #asWord(#buf(N, BUF)) => BUF
 
+    rule #asWord(W0 : W1 : W2 : W3 : #take(28, BUF)) /Int 26959946667150639794667015087019630673637144422540572481103610249216 => #asWord(W0 : W1 : W2 : W3 : .WordStack)
 
+    rule M [ N := BUF ] => M requires #sizeWordStack(BUF) <=Int 0
 
+    rule M [ N := BUF:WordStack ] [ L <- V ]    => M [ L <- V ] [ N := BUF ]                  requires L  <Int N
+    rule M [ N := BUF           ] [ L <- V ]    => M [ L <- V ] [ N := BUF ]                  requires L >=Int N +Int #sizeWordStack(BUF)
+    rule M [ N := BUF           ] [ N <- V ]    => M [ N <- V ] [ N +Int 1 := #drop(1, BUF) ]
```

Comment:
Does this really happen?

Reply:
The scenario is when we get a term like this:

```
M [ 32 := #buf(32, OWNER) ] [ 32 <- 0 ] [ 33 <- 0 ] ... [ 63 <- 0 ]
```

So this, with the rules:

```
rule #drop(N, #drop(M, BUF)) => #drop(N +Int M, BUF)
rule M [ N := BUF ] => M requires #sizeWordStack(BUF) <=Int 0
```

Simplify this term down to:

```
M [ 32 <- 0 ] [ 33 <- 0 ] ... [ 63 <- 0 ]
```

Reply:
hmm, that scenario could be problematic.

i think what happened is that initially we have `... [ 32 := #buf(32, 0) ]`, but it is somehow evaluated to `... [ 32 <- 0 ] [ 33 <- 0 ] ... [ 63 <- 0 ]`. 

it is not incorrect, but in the end, we have `M [ 32 <- 0 ] [ 33 <- 0 ] ... [ 63 <- 0 ]` instead of `M [ 32 := #buf(32, 0) ]`. 

ideally, we want to have `M [ _ := _ ]` form only to allow word-level read/write to local mem, unless we have MSTORE8 opcode in the bytecode.

Reply:
Why is this problematic?

```
M [ 32 := #buf(32, OWNER) ] [ 32 <- 0 ] => M [ 32 <- 0 ] [ 33 := #drop(1, #buf(32, OWNER)) ]
```

is true.

Reply:
The thing is, pushing a `[ 32 <- 0 ]` often allows the update operation to simplify out fully, the the overall resulting term becomes much simpler. So it's easier to debug the proofs and make the final states line up.

Reply:
I'm not saying that this lemma is problematic. I'm saying that you shouldn't met such term `_ [ _ <- _ ]` at the first place.  `_ [ _ := _ ]` is ok, but `_ [ _ <- _ ]` is not.  We'd like to keep `_ [ _ := _ ]` terms, not allowing them to be further evaluated to nested `_ [ _ <- _ ]` terms.  That's the whole point of having `#buf` abstraction, instead of `nthbyteof`.

Reply:
OK. I found what happend. You need to add `EVM-TYPES.mapWriteBytes.recursive` in the concrete-rules.txt file, so that the following rule is not applied for the term `M [ 32 := #buf(32, OWNER) ] [ 32 <- 0 ] [ 33 <- 0 ] ... [ 63 <- 0 ]`:
```
    rule [mapWriteBytes.recursive]: WM[ N := W : WS     ] => (WM[N <- W])[N +Int 1 := WS]
```
And then, you need to add another set of lemmas as I said above.

Reply:
Ah great. I already have the lemmas which handles `M [ N := BUF1 ] [ N := BUF2 ]`, I'll make the correct rule concrete. Thanks!

Reply:
Confirmed that I can remove all the lemmas about `_ [ _ <- _ ]` by marking the appropriate rule concrete. The only remaining rule is:

```
rule M [ N := BUF1 ] [ N := BUF2 ] => M [ N := BUF2 ] requires #sizeWordStack(BUF1) ==Int #sizeWordStack(BUF2)
```

This is enough for now, so I don't think we need to generalize to more cases (where the wordstacks are different sizes for instance) yet.